Update the sql query to delete duplicate sub share

sharidas opened the pull request:
Removal of 'group by id' is required to achieve the
purpose of the repair script. Also made modification
to the query which should support the updated version
of postgresql, mysql.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
`group by id` was causing the results to get skipped and hence returning empty results even if duplicated shares were there. This change would help to solve the issue. 

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
`group by id` was causing the results to get skipped and hence returning empty results even if duplicated shares were there. This change would help to solve the issue.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
Update the sql query to delete duplicate sub share

Removal of 'group by id' is required to achieve the
purpose of the repair script. Also made modification
to the query which should support the updated version
of postgresql, mysql.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Stable 10 branch PR: https://github.com/owncloud/core/pull/31146

PVince81 commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -81,22 +93,13 @@ public function run(IOutput $output) {
+		$rows = $results->fetchAll();
+		$results->closeCursor();
+		foreach ($rows as $row) {
+			$deletedEntries += $this->deleteShareId->setParameter('shareId', (int) $row['id'])
```
would be better to delete in batches as recommended by @butonic instead of one by one

pick the first 1000 entries then pass them in a `delete from oc_share where id in (...)`

sharidas replied:
Implemented by deleting 1000 entries.

PVince81 commented on tests/lib/Repair/RepairSubSharesTest.php:
```diff
@@ -69,6 +69,126 @@ public function deleteAllShares() {
+
+		$usersinsharetable = ['admin', 'user1', 'user2'];
+		foreach ($usersinsharetable as $user) {
+			if ($user === 'admin') {
```
these if statements don't make any sense in the loop for only three users

sharidas replied:
Optimised the code. The code looks better now.

PVince81 commented on tests/lib/Repair/RepairSubSharesTest.php:
```diff
@@ -69,6 +69,126 @@ public function deleteAllShares() {
+			if ($user === 'admin') {
+				$qb->insert('share')
+					->values([
+						'share_type' => $qb->expr()->literal('1'),
```
AFAIK the value should be an integer ?

sharidas replied:
Yes. Updated to integer.

PVince81 commented on tests/lib/Repair/RepairSubSharesTest.php:
```diff
@@ -69,6 +69,126 @@ public function deleteAllShares() {
+		$qb->where($qb->expr()->notIn('id', $qb->createFunction($subQuery->getSQL())));
+		$qb->andWhere($qb->expr()->eq('share_type', $qb->createNamedParameter(2)));
+		$results = $qb->execute()->fetchAll();
+		$this->assertCount(0, $results);
```
please don't duplicate the query from the repair step code, move it to a separate function n the repair step code and call it from here if needed

you only have a few test shares, just store their ids and check whether the ones to be deleted are gone and the others are still there

sharidas replied:
Added new method in the repair step code which would return `getDuplicateRowQueryBuilder()` the query builder.

PVince81 commented on tests/lib/Repair/RepairSubSharesTest.php:
```diff
@@ -69,6 +69,126 @@ public function deleteAllShares() {
+		$qb->select($qb->createFunction('count(*)'))
+			->from('share')
+			->where($qb->expr()->eq('share_with', $qb->createNamedParameter('user2')));
+		$results = $qb->execute()->fetchAll();
```
same here, just check if the share id of that share still exists instead of counting

sharidas replied:
Applied same logic `getDuplicateRowQueryBuilder()` here too.

PVince81 commented on tests/lib/Repair/RepairSubSharesTest.php:
```diff
@@ -131,16 +251,38 @@ public function testPopulateDBAndRemoveDuplicates() {
+		$qb->select('id');
+		$qb->from('share');
+		$qb->where($qb->expr()->notIn('id', $qb->createFunction($subQuery->getSQL())));
+		$qb->andWhere($qb->expr()->eq('share_type', $qb->createNamedParameter(2)));
```
same comments as above

sharidas replied:
Updated the test code to not duplicate the repair code. Also made sure that it verifies the ids that are present and not in the result.

PVince81 replied:
this still looks like the repair code query to me. did you forget pushing ?

sharidas replied:
Yes I missed to push it. Updated. Have to address the other comments.

PVince81 commented on tests/lib/Repair/RepairSubSharesTest.php:
```diff
@@ -185,15 +327,36 @@ public function testLargeDuplicateShareRows() {
+			['share_with' => 'user3', 'parent' => '3'],
+			['share_with' => 'user4', 'parent' => '4'],
+			['share_with' => 'user5', 'parent' => '5'],
+			['share_with' => 'user6', 'parent' => '6'],
```
see, here you checked by known ids..

this will likely not work on Oracle as it sometimes assigns the ids differently. you need to store these from `lastInsertId()` after running your insert queries for the test data

PVince81 reviewed (changes requested):
see above

sharidas commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -55,17 +55,29 @@ public function getName() {
+			->groupBy('share_with')
+			->addGroupBy('parent');
+
+		//$testing = $subQuery->getSQL();
```
Removed.

sharidas commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -55,48 +55,64 @@ public function getName() {
 	}
 
+	public function getDuplicateRowQueryBuilder() {
+		return $this->getDuplicateRows;
```
Made 2 distinct functions 
- `setSelectQueryToDetectDuplicates()` - to set the select query
- `setDeleteShareIds()` - the delete query

sharidas commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -55,48 +55,64 @@ public function getName() {
+				$rows
+			);
+		}
+		//Delete in a batch of 1000 ids
```
Introduced transaction to delete.

sharidas commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -55,48 +55,64 @@ public function getName() {
-			$rows = $results->fetchAll();
+		$results = $this->getDuplicateRows->execute();
+		$rows = $results->fetchAll();
+		$rowId = [];
```
Renamed `$rowId` to `$rowIds`.

jvillafanez commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -48,55 +48,73 @@ public function getName() {
+	 * Set query to delete rows which have duplicate rows of share
+	 * @return IQueryBuilder
+	 */
+	private function setDeleteShareIds() {
```
Change the name to `getDeleteShareIdsBuilder` or something similar. You're *getting* a query builder, not setting anything. In addition, adjust the comment.
Same for the "setSelectQueryToDetectDuplicates" function.

sharidas replied:
Changed the names to
- `getSelectQueryToDetectDuplicatesBuilder()`
- `getDeleteShareIdsBuilder()`
Also updated the comment.

jvillafanez commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -48,55 +48,74 @@ public function getName() {
+		$rows = $results->fetchAll();
+		$rowIds = [];
+		if (\count($rows) > 0) {
 			$results->closeCursor();
```
I guess we want to close this cursor in any case, not just if we have results

sharidas replied:
Moved the close cursor up. So that its closed irrespective of the results.

jvillafanez commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -48,55 +48,74 @@ public function getName() {
+			);
+		}
+		//Delete in a batch of 1000 ids
+		$deleteShareIds->getConnection()->beginTransaction();
```
To improve this:
* Make the deletion conditional. We don't want to bother the DB if there is nothing to be deleted, we also don't want to commit an empty transaction in the same scenario.
* Move the "delete" query builder here, so we don't spend time creating a query we don't need (if there are nothing to be deleted)

sharidas replied:
Done.
Moved the delete query builder here. Also the delete query builder would be called only when there are duplicates.

jvillafanez commented on lib/private/Repair/RepairSubShares.php:
```diff
@@ -48,55 +48,74 @@ public function getName() {
+		$deleteShareIds = $this->getDeleteShareIdsBuilder();
 
-		/**
-		 * Going for pagination because if there are 1 million rows
```
Is this handled somehow, or it something that will never happen? I'm fine if we expect 5000 results at most.

sharidas replied:
Pagination is removed from the code, purposely based on the discussion with @butonic and @PVince81.

jvillafanez commented:
We might need to rollback the transaction if something goes wrong. There is some information in https://stackoverflow.com/questions/10399727/psqlexception-current-transaction-is-aborted-commands-ignored-until-end-of-tra

PVince81 commented:
in general we never called rollback in other repair steps so I don't think we should do so here

jvillafanez commented:
I'm worried because of https://central.owncloud.org/t/upgrade-problem-9-1-6-to-10-0-8-db-error-select-quote-ident/14016

Things go wrong for some strange reason, an exception pops up, but instead of crashing the application (which might be expected) the exception is catched somewhere and the application continues leading to the exception above.

I'm fine if we don't want to rollback the transaction and let the upgrade fail, but we should make sure the upgrade fails and we must know the real reason.

PVince81 commented:
Yes, upgrade should fail on error.

PVince81 commented:
As discussed, @sharidas please confirm that the code will fail on error and no rollback is in place, then this is good to go

jvillafanez commented:
We might want to check if all the repair steps are part of a big transaction. If this is the case, we might want to remove the transaction of these particular repair step, specially if this is the only one. I'd expect that this big transaction is commited after all the repair steps are done.

sharidas commented:
Added line `throw new \Exception()` just before line https://github.com/owncloud/core/pull/31781/files#diff-c7d3960633d9f6fe9eb517281d947301R120. And the the upgrade failed. Below is the output from the console:
```
➜  owncloud4 ./occ maintenance:mode --off              
Cannot load Xdebug - it was already loaded
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Maintenance mode disabled
➜  owncloud4 ./occ upgrade                                                             
Cannot load Xdebug - it was already loaded
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
2018-06-28T12:48:05+00:00 Set log level to debug
2018-06-28T12:48:05+00:00 Turned on maintenance mode
2018-06-28T12:48:05+00:00 Repair step: Repair MySQL database engine
2018-06-28T12:48:05+00:00 Repair info: Not a mysql database -> nothing to do
2018-06-28T12:48:05+00:00 Repair step: Repair MySQL collation
2018-06-28T12:48:05+00:00 Repair info: Not a mysql database -> nothing to do
2018-06-28T12:48:05+00:00 Repair step: Repair SQLite autoincrement
2018-06-28T12:48:05+00:00 Repair step: Repair orphaned reshare
2018-06-28T12:48:05+00:00 Repair step: Repair duplicate entries in oc_lucene_status
2018-06-28T12:48:05+00:00 Repair info: lucene_status table does not exist -> nothing to do
2018-06-28T12:48:05+00:00 Repair step: Upgrade app code from the marketplace
2018-06-28T12:48:05+00:00 Repair info: Using market to update existing apps
2018-06-28T12:48:05+00:00 Repair info: Attempting to update the following missing apps from market: diagnostics
2018-06-28T12:48:05+00:00 Repair info: Fetching app from market: diagnostics
2018-06-28T12:48:13+00:00 Repair info: Attempting to update the following existing compatible apps from market: diagnostics
2018-06-28T12:48:13+00:00 Repair info: Fetching app from market: activity
2018-06-28T12:48:15+00:00 Repair info: 
2018-06-28T12:48:15+00:00 Repair info: Fetching app from market: comments
2018-06-28T12:48:17+00:00 Repair info: App (comments) is not known at the marketplace.
2018-06-28T12:48:17+00:00 Repair info: Fetching app from market: configreport
2018-06-28T12:48:19+00:00 Repair info: App (configreport) is not known at the marketplace.
2018-06-28T12:48:19+00:00 Repair info: Fetching app from market: dav
2018-06-28T12:48:20+00:00 Repair info: App (dav) is not known at the marketplace.
2018-06-28T12:48:20+00:00 Repair info: Fetching app from market: federatedfilesharing
2018-06-28T12:48:22+00:00 Repair info: App (federatedfilesharing) is not known at the marketplace.
2018-06-28T12:48:22+00:00 Repair info: Fetching app from market: federation
2018-06-28T12:48:24+00:00 Repair info: App (federation) is not known at the marketplace.
2018-06-28T12:48:24+00:00 Repair info: Fetching app from market: files
2018-06-28T12:48:26+00:00 Repair info: App (files) is not known at the marketplace.
2018-06-28T12:48:26+00:00 Repair info: Fetching app from market: files_external
2018-06-28T12:48:27+00:00 Repair info: App (files_external) is not known at the marketplace.
2018-06-28T12:48:27+00:00 Repair info: Fetching app from market: files_pdfviewer
2018-06-28T12:48:29+00:00 Repair info: 
2018-06-28T12:48:29+00:00 Repair info: Fetching app from market: files_sharing
2018-06-28T12:48:31+00:00 Repair info: App (files_sharing) is not known at the marketplace.
2018-06-28T12:48:31+00:00 Repair info: Fetching app from market: files_texteditor
2018-06-28T12:48:33+00:00 Repair info: 
2018-06-28T12:48:33+00:00 Repair info: Fetching app from market: files_trashbin
2018-06-28T12:48:34+00:00 Repair info: App (files_trashbin) is not known at the marketplace.
2018-06-28T12:48:34+00:00 Repair info: Fetching app from market: files_versions
2018-06-28T12:48:37+00:00 Repair info: App (files_versions) is not known at the marketplace.
2018-06-28T12:48:37+00:00 Repair info: Fetching app from market: files_videoplayer
2018-06-28T12:48:38+00:00 Repair info: App (files_videoplayer) is not known at the marketplace.
2018-06-28T12:48:38+00:00 Repair info: Fetching app from market: firstrunwizard
2018-06-28T12:48:40+00:00 Repair info: App (firstrunwizard) is not known at the marketplace.
2018-06-28T12:48:40+00:00 Repair info: Fetching app from market: gallery
2018-06-28T12:48:42+00:00 Repair info: 
2018-06-28T12:48:42+00:00 Repair info: Fetching app from market: market
2018-06-28T12:48:44+00:00 Repair info: 
2018-06-28T12:48:44+00:00 Repair info: Fetching app from market: notifications
2018-06-28T12:48:46+00:00 Repair info: App (notifications) is not known at the marketplace.
2018-06-28T12:48:46+00:00 Repair info: Fetching app from market: provisioning_api
2018-06-28T12:48:48+00:00 Repair info: App (provisioning_api) is not known at the marketplace.
2018-06-28T12:48:48+00:00 Repair info: Fetching app from market: systemtags
2018-06-28T12:48:50+00:00 Repair info: App (systemtags) is not known at the marketplace.
2018-06-28T12:48:50+00:00 Repair info: Fetching app from market: templateeditor
2018-06-28T12:48:51+00:00 Repair info: 
2018-06-28T12:48:51+00:00 Repair info: Fetching app from market: updatenotification
2018-06-28T12:48:53+00:00 Repair info: App (updatenotification) is not known at the marketplace.
2018-06-28T12:48:53+00:00 Repair info: App was not updated: activity
2018-06-28T12:48:53+00:00 Repair info: App was not updated: comments
2018-06-28T12:48:53+00:00 Repair info: App was not updated: configreport
2018-06-28T12:48:53+00:00 Repair info: App was not updated: dav
2018-06-28T12:48:53+00:00 Repair info: App was not updated: federatedfilesharing
2018-06-28T12:48:53+00:00 Repair info: App was not updated: federation
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_external
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_pdfviewer
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_sharing
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_texteditor
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_trashbin
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_versions
2018-06-28T12:48:53+00:00 Repair info: App was not updated: files_videoplayer
2018-06-28T12:48:53+00:00 Repair info: App was not updated: firstrunwizard
2018-06-28T12:48:53+00:00 Repair info: App was not updated: gallery
2018-06-28T12:48:53+00:00 Repair info: App was not updated: market
2018-06-28T12:48:53+00:00 Repair info: App was not updated: notifications
2018-06-28T12:48:53+00:00 Repair info: App was not updated: provisioning_api
2018-06-28T12:48:53+00:00 Repair info: App was not updated: systemtags
2018-06-28T12:48:53+00:00 Repair info: App was not updated: templateeditor
2018-06-28T12:48:53+00:00 Repair info: App was not updated: updatenotification
2018-06-28T12:48:53+00:00 Updating database schema
2018-06-28T12:48:53+00:00 Updated database
2018-06-28T12:48:53+00:00 Updating <market> ...
2018-06-28T12:48:53+00:00 Updated <market> to 0.2.4
2018-06-28T12:48:53+00:00 Updating <templateeditor> ...
2018-06-28T12:48:53+00:00 Updated <templateeditor> to 0.3
2018-06-28T12:48:54+00:00 Updating <notifications> ...
2018-06-28T12:48:54+00:00 Updated <notifications> to 0.3.3
2018-06-28T12:48:54+00:00 Repair step: Repair mime types
2018-06-28T12:48:54+00:00 Repair step: Detect file cache entries with path that does not match parent-child relationships
2018-06-28T12:48:54+00:00 Repair step: Generate ETags for file where no ETag is present.
2018-06-28T12:48:54+00:00 Repair info: ETags have been fixed for 0 files/folders.
2018-06-28T12:48:54+00:00 Repair step: Clean tags and favorites
2018-06-28T12:48:54+00:00 Repair info: 0 tags of deleted users have been removed.
2018-06-28T12:48:54+00:00 Repair info: 0 tags for delete files have been removed.
2018-06-28T12:48:54+00:00 Repair info: 0 tag entries for deleted tags have been removed.
2018-06-28T12:48:54+00:00 Repair info: 0 tags with no entries have been removed.
2018-06-28T12:48:54+00:00 Repair step: Drop old database tables
2018-06-28T12:48:54+00:00 Drop old database tables
2018-06-28T12:48:54+00:00 
                                                    2018-06-28T12:48:54+00:00  Done
 28/28 [============================] 100%2018-06-28T12:48:54+00:00 
2018-06-28T12:48:54+00:00 Repair step: Drop old background jobs
2018-06-28T12:48:54+00:00 Repair step: Remove getetag entries in properties table
2018-06-28T12:48:54+00:00 Repair info: Removed 0 unneeded "{DAV:}getetag" entries from properties table.
2018-06-28T12:48:54+00:00 Repair step: Repair outdated OCS IDs
2018-06-28T12:48:54+00:00 Repair step: Repair invalid shares
2018-06-28T12:48:54+00:00 Repair step: Remove old share propagation app entries
2018-06-28T12:48:54+00:00 Repair step: Move user avatars outside the homes to the new location
2018-06-28T12:48:54+00:00 Repair step: Remove shares of a users root folder
2018-06-28T12:48:54+00:00 Repair step: Repair unmerged shares
2018-06-28T12:48:54+00:00 Repair step: Disable extra themes
2018-06-28T12:48:54+00:00 Repair step: Repair sub shares
2018-06-28T12:48:54+00:00 Exception: 
2018-06-28T12:48:54+00:00 Update failed
2018-06-28T12:48:54+00:00 Maintenance mode is kept active
2018-06-28T12:48:54+00:00 Reset log level
➜  owncloud4
```

PVince81 commented:
AFAIK there is no transaction wrapping all the repair steps, that would be a loooot of data considering the heavy operations that some steps might do.

in case of breakage, the admin must restore the DB from backup

PVince81 commented:
also note that some repair steps might touch files on disk, so even if we'd revert transactions globally, we cannot guarantee atomicity across DB and disk.

@jvillafanez are you ok for merging ?

PVince81 commented:
what I meant above is that we don't have a big transaction that wraps all repair steps.

there are only two repair steps currently using transactions: RepairMismatchFileCache, but this code is driven by occ files:scan --repair also. The other one is SqliteAutoincrement.

So yes, maybe just remove the transaction then.

PVince81 commented:
@sharidas please remove the transaction for consistency

sharidas commented:
Retested again with the change made with the scenario https://github.com/owncloud/core/pull/30695#issuecomment-381689933:

```
➜  owncloud4 ./occ upgrade              
Cannot load Xdebug - it was already loaded
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
2018-07-04T11:59:55+00:00 Set log level to debug
2018-07-04T11:59:55+00:00 Repair step: Repair MySQL database engine
2018-07-04T11:59:55+00:00 Repair info: Not a mysql database -> nothing to do
2018-07-04T11:59:55+00:00 Repair step: Repair MySQL collation
2018-07-04T11:59:55+00:00 Repair info: Not a mysql database -> nothing to do
2018-07-04T11:59:55+00:00 Repair step: Repair SQLite autoincrement
2018-07-04T11:59:55+00:00 Repair step: Repair orphaned reshare
2018-07-04T11:59:55+00:00 Repair step: Repair duplicate entries in oc_lucene_status
2018-07-04T11:59:55+00:00 Repair info: lucene_status table does not exist -> nothing to do
2018-07-04T11:59:55+00:00 Repair step: Upgrade app code from the marketplace
2018-07-04T11:59:55+00:00 Repair info: Using market to update existing apps
2018-07-04T11:59:55+00:00 Repair info: Attempting to update the following missing apps from market: diagnostics
2018-07-04T11:59:55+00:00 Repair info: Fetching app from market: diagnostics
2018-07-04T12:00:06+00:00 Repair info: Attempting to update the following existing compatible apps from market: diagnostics
2018-07-04T12:00:06+00:00 Repair info: Fetching app from market: activity
2018-07-04T12:00:10+00:00 Repair info: 
2018-07-04T12:00:10+00:00 Repair info: Fetching app from market: comments
2018-07-04T12:00:15+00:00 Repair info: App (comments) is not known at the marketplace.
2018-07-04T12:00:15+00:00 Repair info: Fetching app from market: configreport
2018-07-04T12:00:19+00:00 Repair info: App (configreport) is not known at the marketplace.
2018-07-04T12:00:19+00:00 Repair info: Fetching app from market: dav
2018-07-04T12:00:23+00:00 Repair info: App (dav) is not known at the marketplace.
2018-07-04T12:00:23+00:00 Repair info: Fetching app from market: federatedfilesharing
2018-07-04T12:00:27+00:00 Repair info: App (federatedfilesharing) is not known at the marketplace.
2018-07-04T12:00:27+00:00 Repair info: Fetching app from market: federation
2018-07-04T12:00:30+00:00 Repair info: App (federation) is not known at the marketplace.
2018-07-04T12:00:30+00:00 Repair info: Fetching app from market: files
2018-07-04T12:00:34+00:00 Repair info: App (files) is not known at the marketplace.
2018-07-04T12:00:34+00:00 Repair info: Fetching app from market: files_external
2018-07-04T12:00:38+00:00 Repair info: App (files_external) is not known at the marketplace.
2018-07-04T12:00:38+00:00 Repair info: Fetching app from market: files_pdfviewer
2018-07-04T12:00:41+00:00 Repair info: 
2018-07-04T12:00:41+00:00 Repair info: Fetching app from market: files_sharing
2018-07-04T12:00:45+00:00 Repair info: App (files_sharing) is not known at the marketplace.
2018-07-04T12:00:45+00:00 Repair info: Fetching app from market: files_texteditor
2018-07-04T12:00:49+00:00 Repair info: 
2018-07-04T12:00:49+00:00 Repair info: Fetching app from market: files_trashbin
2018-07-04T12:00:52+00:00 Repair info: App (files_trashbin) is not known at the marketplace.
2018-07-04T12:00:52+00:00 Repair info: Fetching app from market: files_versions
2018-07-04T12:00:55+00:00 Repair info: App (files_versions) is not known at the marketplace.
2018-07-04T12:00:55+00:00 Repair info: Fetching app from market: files_videoplayer
2018-07-04T12:00:57+00:00 Repair info: App (files_videoplayer) is not known at the marketplace.
2018-07-04T12:00:57+00:00 Repair info: Fetching app from market: firstrunwizard
2018-07-04T12:00:59+00:00 Repair info: App (firstrunwizard) is not known at the marketplace.
2018-07-04T12:00:59+00:00 Repair info: Fetching app from market: gallery
2018-07-04T12:01:01+00:00 Repair info: 
2018-07-04T12:01:01+00:00 Repair info: Fetching app from market: market
2018-07-04T12:01:03+00:00 Repair info: 
2018-07-04T12:01:03+00:00 Repair info: Fetching app from market: notifications
2018-07-04T12:01:05+00:00 Repair info: App (notifications) is not known at the marketplace.
2018-07-04T12:01:05+00:00 Repair info: Fetching app from market: provisioning_api
2018-07-04T12:01:07+00:00 Repair info: App (provisioning_api) is not known at the marketplace.
2018-07-04T12:01:07+00:00 Repair info: Fetching app from market: systemtags
2018-07-04T12:01:09+00:00 Repair info: App (systemtags) is not known at the marketplace.
2018-07-04T12:01:09+00:00 Repair info: Fetching app from market: templateeditor
2018-07-04T12:01:11+00:00 Repair info: 
2018-07-04T12:01:11+00:00 Repair info: Fetching app from market: updatenotification
2018-07-04T12:01:13+00:00 Repair info: App (updatenotification) is not known at the marketplace.
2018-07-04T12:01:13+00:00 Repair info: App was not updated: activity
2018-07-04T12:01:13+00:00 Repair info: App was not updated: comments
2018-07-04T12:01:13+00:00 Repair info: App was not updated: configreport
2018-07-04T12:01:13+00:00 Repair info: App was not updated: dav
2018-07-04T12:01:13+00:00 Repair info: App was not updated: federatedfilesharing
2018-07-04T12:01:13+00:00 Repair info: App was not updated: federation
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_external
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_pdfviewer
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_sharing
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_texteditor
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_trashbin
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_versions
2018-07-04T12:01:13+00:00 Repair info: App was not updated: files_videoplayer
2018-07-04T12:01:13+00:00 Repair info: App was not updated: firstrunwizard
2018-07-04T12:01:13+00:00 Repair info: App was not updated: gallery
2018-07-04T12:01:13+00:00 Repair info: App was not updated: market
2018-07-04T12:01:13+00:00 Repair info: App was not updated: notifications
2018-07-04T12:01:13+00:00 Repair info: App was not updated: provisioning_api
2018-07-04T12:01:13+00:00 Repair info: App was not updated: systemtags
2018-07-04T12:01:13+00:00 Repair info: App was not updated: templateeditor
2018-07-04T12:01:13+00:00 Repair info: App was not updated: updatenotification
2018-07-04T12:01:13+00:00 Updating database schema
2018-07-04T12:01:13+00:00 Updated database
2018-07-04T12:01:13+00:00 Updating <market> ...
2018-07-04T12:01:13+00:00 Updated <market> to 0.2.4
2018-07-04T12:01:13+00:00 Updating <templateeditor> ...
2018-07-04T12:01:13+00:00 Updated <templateeditor> to 0.3
2018-07-04T12:01:13+00:00 Updating <notifications> ...
2018-07-04T12:01:13+00:00 Updated <notifications> to 0.3.3
2018-07-04T12:01:13+00:00 Repair step: Repair mime types
2018-07-04T12:01:13+00:00 Repair step: Detect file cache entries with path that does not match parent-child relationships
2018-07-04T12:01:13+00:00 Repair step: Generate ETags for file where no ETag is present.
2018-07-04T12:01:13+00:00 Repair info: ETags have been fixed for 0 files/folders.
2018-07-04T12:01:13+00:00 Repair step: Clean tags and favorites
2018-07-04T12:01:13+00:00 Repair info: 0 tags of deleted users have been removed.
2018-07-04T12:01:13+00:00 Repair info: 0 tags for delete files have been removed.
2018-07-04T12:01:13+00:00 Repair info: 0 tag entries for deleted tags have been removed.
2018-07-04T12:01:13+00:00 Repair info: 0 tags with no entries have been removed.
2018-07-04T12:01:13+00:00 Repair step: Drop old database tables
2018-07-04T12:01:13+00:00 Drop old database tables
2018-07-04T12:01:13+00:00 
                                                    2018-07-04T12:01:13+00:00  Done
 28/28 [============================] 100%2018-07-04T12:01:13+00:00 
2018-07-04T12:01:13+00:00 Repair step: Drop old background jobs
2018-07-04T12:01:13+00:00 Repair step: Remove getetag entries in properties table
2018-07-04T12:01:13+00:00 Repair info: Removed 0 unneeded "{DAV:}getetag" entries from properties table.
2018-07-04T12:01:13+00:00 Repair step: Repair outdated OCS IDs
2018-07-04T12:01:13+00:00 Repair step: Repair invalid shares
2018-07-04T12:01:13+00:00 Repair step: Remove old share propagation app entries
2018-07-04T12:01:13+00:00 Repair step: Move user avatars outside the homes to the new location
2018-07-04T12:01:13+00:00 Repair step: Remove shares of a users root folder
2018-07-04T12:01:13+00:00 Repair step: Repair unmerged shares
2018-07-04T12:01:13+00:00 Repair step: Disable extra themes
2018-07-04T12:01:13+00:00 Repair step: Repair sub shares
2018-07-04T12:01:13+00:00 Repair info: Removed 1 shares where duplicate rows where found
2018-07-04T12:01:13+00:00 Starting code integrity check...
2018-07-04T12:01:27+00:00 Finished code integrity check
2018-07-04T12:01:27+00:00 Update successful
2018-07-04T12:01:27+00:00 Maintenance mode is kept active
2018-07-04T12:01:27+00:00 Reset log level
➜  owncloud4
```

Below is the data in the oc_share table:
```
sqlite> select * from oc_share;
id          share_type  share_with  uid_owner   uid_initiator  parent      item_type   item_source  item_target  file_target  permissions  stime       accepted    expiration  token       mail_send   share_name  file_source
----------  ----------  ----------  ----------  -------------  ----------  ----------  -----------  -----------  -----------  -----------  ----------  ----------  ----------  ----------  ----------  ----------  -----------
1           1           group1      admin       admin                      folder      25                        /test        31           1530705313  0                                   0                       25         
2           2           user1       admin       admin          1           folder      25                        /test        0            1530705313  0                                   0                       25         
3           2           user2       admin       admin          1           folder      25                        /test_mine   31           1530705313  0                                   0                       25         
sqlite> 
```

PVince81 reviewed (approved):
:+1:

please adjust the backport
